src/hotspot/share/oops/method.cpp

```diff
@@ -141,10 +141,9 @@ void Method::deallocate_contents(ClassLoaderData* loader_data) {
 
 void Method::release_C_heap_structures() {
   if (method_data()) {
-#if INCLUDE_JVMCI
-    FailedSpeculation::free_failed_speculations(method_data()->get_failed_speculations_address());
-#endif
-    // Destroy MethodData
+    method_data()->release_C_heap_structures();
+
+    // Destroy MethodData embedded lock
     method_data()->~MethodData();
```

Comment:
I like this approach, but this direct call to the destructor still being here is odd to me. Is that intentional? Is it possible for this code path to execute and then eventually MethodFactory::free_metadata for MethodData? Because that will cause the destructor to be called twice.

Reply:
It's intentional.  No, this code path is only called once when we unload the classes.  There are two code paths - one with unloading and one with deallocation.  The latter was for redefinition, class unloading failure, a couple of other class loading things, and now this MethodData initialization path.